SwiftFlow/CoreTypes/Store.swift

```diff
@@ -1,216 +1,135 @@
-     return type, e.g. to return promises
-     */
-    func dispatch(actionCreator: ActionCreator) -> Any
-
-    /**
-     Dispatches an async action creator to the store. An async action creator generates an
-     action creator asynchronously.
-     */
-    func dispatch(asyncActionCreator: AsyncActionCreator)
-
-    /**
-     Dispatches an action and calls the callback as soon as the action has been processed.
-     You will receive the updated store state as part of this callback.
-
-     Example of dispatching an action and implementing a callback:
-
-     ```
-     store.dispatch( CounterAction.IncreaseCounter ) { state in
-        print("New state: \(state)")
-     }
-     ```
-
-     - parameter action: The action that is being dispatched to the store
-     - returns: By default returns the dispatched action, but middlewares can change the
-     return type, e.g. to return promises
-     */
-    func dispatch(action: Action, callback: DispatchCallback?) -> Any
-
-    /**
-     Dispatches an action creator to the store. Action creators are functions that generate
-     actions. They are called by the store and receive the current state of the application
-     and a reference to the store as their input.
-
-     Based on that input the action creator can either return an action or not. Alternatively
-     the action creator can also perform an asynchronous operation and dispatch a new action
-     at the end of it.
-
-     Example of an action creator:
-
-     ```
-     func deleteNote(noteID: Int) -> ActionCreator {
-        return { state, store in
-            // only delete note if editing is enabled
-            if (state.editingEnabled == true) {
-                return NoteDataAction.DeleteNote(noteID)
-            } else {
-                return nil
-            }
+    }
+
+    public var dispatchFunction: DispatchFunction!
+
+    private var reducer: AnyReducer
+    var subscribers: [AnyStoreSubscriber] = []
+    private var isDispatching = false
+
+    public required convenience init(reducer: AnyReducer, state: State?) {
+        self.init(reducer: reducer, state: state, middleware: [])
+    }
+
+    public required init(reducer: AnyReducer, state: State?, middleware: [Middleware]) {
+        self.reducer = reducer
+
+        // Wrap the dispatch function with all middlewares
+        self.dispatchFunction = middleware.reverse().reduce(self._defaultDispatch) {
+            dispatchFunction, middleware in
+            return middleware(self.dispatch, { self.state })(dispatchFunction)
         }
-     }
-     ```
 
-     This action creator can then be dispatched as following:
+        if let state = state {
+            self.state = state
+        } else {
+            dispatch(SwiftFlowInit())
+        }
+    }
 
-     ```
-     store.dispatch( noteActionCreatore.deleteNote(3) )
-     ```
+    public func subscribe(subscriber: AnyStoreSubscriber) {
+        guard subscribers.indexOf({ $0 === subscriber }) == nil else {
```

Comment:
Is there a reason this is preferred over
```swift
if subscribers.contains({ $0 === subscriber }) { ... }
```
I believe it's a bit more readable and I feel like you'd use indexOf only when you plan to use the resulting index